adds `:variable` parameter substitution

edublancas opened the pull request:
## Describe your changes

also fixing flake8 lint issues since version 6.1.0 added new rules

## Issue number

Closes #671

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [ ] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [ ] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)



<!-- readthedocs-preview jupysql start -->
----
:books: Documentation preview :books:: https://jupysql--770.org.readthedocs.build/en/770/

<!-- readthedocs-preview jupysql end -->

neelasha23 commented on doc/user-guide/template.md:
````diff
@@ -42,6 +42,11 @@ Let's load some data and connect to the in-memory DuckDB instance:
 ```{code-cell} ipython3
 %load_ext sql
+%sql duckdb://
+%config SqlMagic.displaylimit = 3
````
We need to add `displaylimit` [here](https://jupysql.ploomber.io/en/latest/api/configuration.html#) also

edublancas replied:
good catch. added https://github.com/ploomber/jupysql/pull/770/commits/8d07db41ba3e4a3b79117b9a59e48681732a1ae0

neelasha23 commented:
Observations:

1. Not directly related to the issue, but loading a dataset is printing this message `Found pyproject.toml from '/Users/neelashasen/Dev/jupysql_master/jupysql'`. ... It's not very clear what's the purpose because no settings are configured. Maybe we can add a message for the user to understand how to use the `pyproject.toml` (`You may add configuration settings <link to [configurations](https://jupysql.ploomber.io/en/latest/api/configuration.html) )

2. When trying to use `:variable` without setting `%config SqlMagic.named_paramstyle = True` it prints entire stacktrace: 
```
 1061     else:
-> 1062         raise exc.InvalidRequestError(
   1063             "A value is required for bind parameter %r"
   1064             % bindparam.key,
   1065             code="cd3x",
   1066         )
   1068 if resolved_extracted:
   1069     value_param = resolved_extracted.get(bindparam, bindparam)

StatementError: (sqlalchemy.exc.InvalidRequestError) A value is required for bind parameter 'sex'
[SQL: SELECT *
FROM penguins.csv
WHERE sex = ?]
(Background on this error at: https://sqlalche.me/e/14/cd3x)
```

Maybe we can raise some custom exception and direct user to enable `named_paramstyle`.

neelasha23 commented on src/tests/test_parse.py:
```diff
@@ -278,3 +279,30 @@ def test_magic_args(ip, line, expected):
+        ),
+        (
+            'SELECT * FROM table where x > "":x""',
+            'SELECT * FROM table where x > ""\\:x""',
```
<img width="736" alt="Screenshot 2023-07-30 at 7 35 24 PM" src="https://github.com/ploomber/jupysql/assets/29096135/518c24b6-31b0-4d93-8b46-974b45e0dd5d">

Is this invalid usecase? It looks similar to `"SELECT * FROM table where x > ':x'"` ?

Also, what about `"SELECT * FROM table where x > "':x'" "` ?

edublancas replied:
yes, they're are invalid, named parameters should not be quoted, otherwise they're considered string literals. I added a new warning: if named_paramstyle is True, the query contains quoted params and there is a defined variable with the same name, I'm showing a warning telling the user that if they want to use parameters, they need to unquote them

edublancas commented:
> Observations:
> 
>     1. Not directly related to the issue, but loading a dataset is printing this message `Found pyproject.toml from '/Users/neelashasen/Dev/jupysql_master/jupysql'`. ... It's not very clear what's the purpose because no settings are configured. Maybe we can add a message for the user to understand how to use the `pyproject.toml` (`You may add configuration settings <link to [configurations](https://jupysql.ploomber.io/en/latest/api/configuration.html)> )


opened https://github.com/ploomber/jupysql/issues/771

> 
> <img alt="Screenshot 2023-07-30 at 7 20 14 PM" width="819" src="https://user-images.githubusercontent.com/29096135/257055463-1fbe2ae4-00b4-42af-994f-0c0e831f8454.png">
> 
>     3. When trying to use `:variable` without setting `%config SqlMagic.named_paramstyle = True` it prints entire stacktrace:
> 
> 
> ```
>  1061     else:
> -> 1062         raise exc.InvalidRequestError(
>    1063             "A value is required for bind parameter %r"
>    1064             % bindparam.key,
>    1065             code="cd3x",
>    1066         )
>    1068 if resolved_extracted:
>    1069     value_param = resolved_extracted.get(bindparam, bindparam)
> 
> StatementError: (sqlalchemy.exc.InvalidRequestError) A value is required for bind parameter 'sex'
> [SQL: SELECT *
> FROM penguins.csv
> WHERE sex = ?]
> (Background on this error at: https://sqlalche.me/e/14/cd3x)
> ```
> 
> Maybe we can raise some custom exception and direct user to enable `named_paramstyle`.

fixed this


>     3. This is a valid testcase, we can add it :


> 
> <img alt="Screenshot 2023-07-30 at 7 48 39 PM" width="833" src="https://user-images.githubusercontent.com/29096135/257056656-f79cf065-06bf-4fcc-9222-229de7098b53.png">

this is db-dependent. the database is casting your parameter to the right type, but it won't work across al databases, users need to ensure the variable has the right type


>     4. Do we want to handle this in another PR? (detect that user is trying to use a query parameter but has not declared it)
> 
> 
> <img alt="Screenshot 2023-07-30 at 7 49 48 PM" width="634" src="https://user-images.githubusercontent.com/29096135/257056716-e94b85a9-0c8d-4e29-86e5-6c57315cc53a.png">


this isn't a valid case `':param'` is interpreted as a literal because it's quoted. but I added a warning, if they do this and they have a param variable defined, they'll see the warning saying they need to unquote

neelasha23 reviewed (approved):
looks good!
